File: PopulatingNextRightPointersinEachNode.cpp
#include "Utility.h"

class Solution 
{
public:
    void connect(TreeLinkNode *root) 
    {
    	if (root == NULL)
    		return ;
    	queue<TreeLinkNode*> q;
    	queue<int> l;
    	TreeLinkNode *t;
    	int curlevel;
    	q.push(root);
    	l.push(1);
    	while (!q.empty())
    	{
    		t = q.front();
    		q.pop();
    		int curlevel = l.front();
    		l.pop();
    		if (l.size() == 0 || l.front() > curlevel)
    		{
    			t->next = NULL;
    		}
    		else
    		{
    			t->next = q.front();
    		}
    		if (t->left != NULL)
    		{
    			q.push(t->left);
    			l.push(curlevel + 1);
    		}
    		if (t->right != NULL)
    		{
    			q.push(t->right);
    			l.push(curlevel + 1);
    		}
    	}
    }
};

int main()
{
	return 0;
}

File: JavaSolution/src/PopulatingNextRightPointersinEachNode.java
/**
 * Created by xyunpeng on 1/27/16.
 */
public class PopulatingNextRightPointersinEachNode {
    public void connect(TreeLinkNode root) {
        if (root == null) {
            return ;
        }

        TreeLinkNode level = root;
        while (level.left != null) {
            TreeLinkNode p = level;
            while (p != null) {
                p.left.next = p.right;
                if (p.next != null) p.right.next = p.next.left;
                p = p.next;
            }
            level = level.left;
        }
    }
}
